### PR TITLE
[On-chain, Relayminer] chore: add claim msg validation

### PR DESCRIPTION
**Pull request by bryanchriswhite**
### Human Summary

Adds validation logic around the claim's session during claim creation. The session ID must match that as determined by an on-chain query of the session given the same inputs. Additionally, the supplier list for the session must contain the supplier whose address is referenced in the claim.

### Summary generated by Reviewpad on 07 Dec 23 10:58 UTC
This pull request includes various changes across multiple files. Here is a summary of the changes:

- `supplier.go` (in `testutil/keeper` package):
  - Added imports for `supplier` and `supplier/mocks` packages.
  - Added import for `sessiontypes` package from `github.com/pokt-network/poktroll/x/session/types`.
  - Modified the `SupplierKeeper` function to accept an additional parameter `sessionByAppAddr` of type `supplier.SessionsByAppAddress`.
  - Added implementation for `GetSession` method of the `mockSessionKeeper` mock.
  - Added a call to `k.SupplySessionKeeper(mockSessionKeeper)`.

- `relay_verifier.go` (in `pkg/relayer/proxy` package):
  - Added import for `sdkerrors` package.
  - Added import for `ring_secp256k1` package.
  - Removed import for `github.com/cometbft/cometbft/crypto` package.
  - Removed import for `github.com/pokt-network/poktroll/x/session/types` package.
  - Modified the `VerifyRelayRequest` function to handle missing application address from the relay request.
  - Changed the method of obtaining signable bytes of the relay request to `GetSignableBytesHash`.
  - Changed the method of verifying relay request's signature using `ringSig`.
  - Modified the code for verifying relay request session using `GetSession` method of `rp.sessionQuerier`.
  - Improved error message when session ID does not match relay request's session ID.

- `relay_builders.go`:
  - Added a debug log message when there is a failure in unmarshaling the relay request.

- `ring_signer.go`:
  - Changed the `Sign` function to accept an array of 32 bytes for the message argument instead of a byte slice.

- `sessiontypes` package:
  - Added `SessionKeeper` interface which includes the `GetSession` method.
  - Updated the `go:generate mockgen` command to include the `SessionKeeper` interface in the mock generation.

- `testqueryclients/appquerier.go`:
  - Updated the import paths for the `mockclient` and `types` packages.
  - Replaced the context parameter in the `GetApplication` function with an underscore.
  - Updated the return type of the `GetApplication` function to use the `apptypes` type.
  - Updated the variable names within the `GetApplication` function.
  - Updated the error type returned in case of an application not found to `apptypes.ErrAppNotFound`.

- `app.go`:
  - Changes to the `New` function related to the ordering of code and addition of comments.

- `interface.go`:
  - Added `SupplierQueryClient` interface.
  - Added `SessionQueryClient` interface.

- `session.feature`:
  - Added a new feature file describing the "Session Namespace" feature.
  - Includes a scenario for completing the claim/proof lifecycle for a valid session.

- `simple_signer.go`:
  - Changed the signature parameter of the `Sign` function to accept an array of bytes with a fixed size of 32.

- `logger.go` (in `pkg/polylog/polyzero` package):
  - Added the `os` package as an import.
  - Updated the `NewLogger` function to provide default logger configuration.
  - Added TODO comments for improving and adding new logger functions.

- `client.go`:
  - Added import of the `cosmossdk.io/api/tendermint/abci` package.
  - Removed import of the `abciTypes` package from `github.com/cometbft/cometbft/abci/types`.
  - Changed the type `TxEvent` to be an alias for the `abci.TxResult` type.
  - Added comments.

- `proxy.go`:
  - Removed the import of `cosmosclient` package.
  - Removed the import of `querytypes` package.
  - Replaced the type `suppliertypes.QueryClient` with `client.SupplierQueryClient`.
  - Replaced the type `sessiontypes.QueryClient` with `client.SessionQueryClient`.
  - Removed the field `clientCtx`.
  - Replaced the initialization of `supplierQuerier` with `client.SupplierQueryClient`.
  - Replaced the initialization of `sessionQuerier` with `client.SessionQueryClient`.
  - Removed the initialization of `keyring`.
  - Added a check to ensure `proxiedServicesEndpoints` is not empty in the `validateConfig` method.

- `relayerproxy.go`:
  - Added various package imports and new struct and functions related to testing utilities and mock structures for the relayer proxy.

## Issue

- #140 
- #237 

On-chain claim message handling should validate that the claimed session matches the session at the given claimed sesison end height; i.e. given some session inputs, ensure that the ID matches and that the supplier is in the session.

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Run all unit tests**: `make go_develop_and_test`
- [ ] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: testutil/keeper/supplier.go [hunk 1]
```diff
@@ -38,6 +50,40 @@ func SupplierKeeper(t testing.TB) (*keeper.Keeper, sdk.Context) {
...
+						Header: &sessiontypes.SessionHeader{
+							ApplicationAddress:      sessionMock.AppAddr,
+							Service:                 req.GetService(),
+							SessionStartBlockHeight: 1,
+							SessionId:               sessionMock.SessionId,
+							SessionEndBlockHeight:   5,
+						},
+						SessionId:           sessionMock.SessionId,
+						SessionNumber:       1,
+						NumBlocksPerSession: 4,
```
Comment (bryanchriswhite) [inline testutil/keeper/supplier.go:75]: Where would these integers be better sourced from?

**Review by Olshansk — CHANGES_REQUESTED** [review]
Posting a partial review and will finish tomorrow.

File: e2e/tests/session.feature [hunk 2]
```diff
@@ -0,0 +1,9 @@
+Feature: Session Namespace
+
+  Scenario: Supplier completes claim/proof lifecycle for a valid session
+    Given the user has the pocketd binary installed
+    And the supplier "supplier1" has serviced a session of relays for application "app1"
+    When after the supplier creates a claim for the session
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply e2e/tests/session.feature:6]: I think the "when" should line up with the action, which I think is actually the previous step. I also think it makes sense for the action to be in present tense. Then, with an "and" in front instead, "after..." reads much  better IMO. I feel like the "after" is a really clear way to communicate that this step will be waiting for something.

```gherkin
Scenario: Supplier completes claim/proof lifecycle for a valid session
    Given the user has the pocketd binary installed
    When the supplier "supplier1" services a session with "5" relays for service "anvil" for application "app1"
    And after the supplier creates a claim for the session
```

File: e2e/tests/session_steps_test.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+func (s *suite) TheSupplierHasServicedASessionOfRelaysForApplication(supplierName string, appName string) {
+	// TODO_IN_THIS_COMMIT: use consts or something
+	pocketNodeWebsocketUrl := "ws://localhost:36657/websocket"
+	msgClaimSenderQueryFmt := "tm.event='Tx' AND message.sender='%s'"
+	msgSenderQuery := fmt.Sprintf(msgClaimSenderQueryFmt, accNameToAddrMap[supplierName])
+	s.Logf("msgSenderQuery: %s", msgSenderQuery)
+	ctx := context.Background()
+
+	// TODO_TECHDEBT: refactor to use EventsReplayClient once available.
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply e2e/tests/session_steps_test.go:116]: Yes, #220

File: testutil/keeper/supplier.go (hunk 1 above)
Comment (bryanchriswhite) [reply testutil/keeper/supplier.go:75]: Sorry, I'm not following. Can you elaborate?

File: e2e/tests/session_steps_test.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	//require.Lenf(s, claimsRes.Claim, 1, "expected 1 claim, got %d", len(claimsRes.Claim))
+	require.NotEmpty(s, claimsRes.Claim)
+
+	claim := claimsRes.Claim[0]
+	require.Equal(s, accNameToAddrMap[supplierName], claim.SupplierAddress)
+}
+
+func (s *suite) TheSupplierHasServicedASessionOfRelaysForApplication(supplierName string, appName string) {
+	// TODO_IN_THIS_COMMIT: use consts or something
+	pocketNodeWebsocketUrl := "ws://pocket-sequencer:36657/websocket"
```
Comment (bryanchriswhite) [inline e2e/tests/session_steps_test.go:110]:
```suggestion
	// TODO_IN_THIS_COMMIT: use testclient.CometLocalWebsocketURL instead.
	pocketNodeWebsocketUrl := "ws://poktroll-sequencer:36657/websocket"
```

File: x/session/keeper/session_hydrator.go [hunk 5]
```diff
@@ -157,7 +157,11 @@ func (k Keeper) hydrateSessionSuppliers(ctx sdk.Context, sh *sessionHydrator) er
 	suppliers := k.supplierKeeper.GetAllSupplier(ctx)
 
 	candidateSuppliers := make([]*sharedtypes.Supplier, 0)
-	for _, supplier := range suppliers {
+	for _, s := range suppliers {
+		// NB: Allocate a new heap variable as s is a value and we're appending
```
Comment (Olshansk) [inline x/session/keeper/session_hydrator.go:161]: Ty for catching htis!

File: x/supplier/client/cli/query_claim_test.go [hunk 6]
```diff
@@ -84,64 +95,151 @@ func createClaim(
...
 	t.Helper()
 
-	// Prepare the network
+	// Initialize a network config.
 	cfg := network.DefaultConfig()
+
+	// Construct an in-memory keyring so that it can be populated and used prior
+	// to network start.
+	kr := keyring.NewInMemory(cfg.Codec)
+	// Populate the in-memmory keyring with as many pre-generated accounts as
```
Comment (Olshansk) [inline x/supplier/client/cli/query_claim_test.go:148]: `as we expect to need for the test`

Can `20` be a function of all the counts provided to the function?

File: x/supplier/client/cli/query_claim_test.go [hunk 7]
```diff
@@ -84,64 +95,151 @@ func createClaim(
...
+	for sessionIdx := 0; sessionIdx < sessionCount; sessionIdx++ {
 		sessionEndHeight += numBlocksPerSession
-		sessionId := fmt.Sprintf("session_id%d", sessionNum)
-		for claimNum := 0; claimNum < numClaimsPerSession; claimNum++ {
-			supplierAddr := addresses[claimNum]
-			claim := createClaim(t, net, ctx, supplierAddr, sessionId, sessionEndHeight)
-			claims = append(claims, *claim)
-			// TODO_TECHDEBT(#196): Move this outside of the forloop so that the test iteration is faster
-			require.NoError(t, net.WaitForNextBlock())
+		// TODO_IN_THIS_COMMIT: numClaimsPerSession == supplierCount * appCount
```
Comment (Olshansk) [inline x/supplier/client/cli/query_claim_test.go:211]: Weird that the bot didn't pick this up

File: x/supplier/client/cli/query_claim_test.go [hunk 8]
```diff
@@ -84,64 +95,151 @@ func createClaim(
...
+		for _, supplierAcct := range supplierAccts {
```
Comment (Olshansk) [inline x/supplier/client/cli/query_claim_test.go:212]: NIT: loop through apps on the outside.

Fundamentally it results in the same thing, but logically it makes more sense given the `1 app -> N suppliers` relationship in a session.

File: x/supplier/client/cli/query_claim_test.go [hunk 9]
```diff
@@ -84,64 +95,151 @@ func createClaim(
...
+			for _, appAcct := range appAccts {
+				claim := createClaim(
+					t, net, ctx,
+					supplierAcct.Address.String(),
+					sessionEndHeight,
+					appAcct.Address.String(),
+				)
+				claims = append(claims, *claim)
+				// TODO_TECHDEBT(#196): Move this outside of the forloop so that the test iteration is faster
```
Comment (Olshansk) [inline x/supplier/client/cli/query_claim_test.go:221]: You have not idea how much time I wasted on this...

File: x/supplier/client/cli/query_claim_test.go [hunk 10]
```diff
@@ -84,64 +95,151 @@ func createClaim(
...
 func TestClaim_Show(t *testing.T) {
-	numSessions := 1
-	numClaimsPerSession := 2
-
-	net, claims := networkWithClaimObjects(t, numSessions, numClaimsPerSession)
+	sessionCount := 1
+	supplierCount := 3
+	appCount := 3
+	serviceCount := 1
+	//numClaimsPerSession := supplierCount * appCount * serviceCount
```
Comment (Olshansk) [inline x/supplier/client/cli/query_claim_test.go:235]: TODO_IN_THIS_PR?

File: x/supplier/keeper/msg_server_create_claim_test.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+	"github.com/pokt-network/poktroll/x/supplier/keeper"
+	"github.com/pokt-network/poktroll/x/supplier/types"
+	suppliertypes "github.com/pokt-network/poktroll/x/supplier/types"
+)
+
+func TestMsgServer_CreateClaim_Success(t *testing.T) {
+	appAddr, supplierAddr := sample.AccAddress(), sample.AccAddress()
+
+	// TODO_IN_THIS_COMMIT: dedup & refactor to a test helper.
+	sessionMockMap := keepertest.SessionMetaFixturesByAppAddr{
```
Comment (Olshansk) [inline x/supplier/keeper/msg_server_create_claim_test.go:22]: Per one of my other comments. Reconsider: s/mock/fixture or the use of this structure at all.

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: e2e/tests/session.feature [hunk 12]
```diff
@@ -0,0 +1,9 @@
+Feature: Session Namespace
+
+  Scenario: Supplier completes claim/proof lifecycle for a valid session
+    Given the user has the pocketd binary installed
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply e2e/tests/session.feature:4]: Would you mind if we do  this in a separate PR as it will require updating all existing feature files as well.

File: testutil/keeper/supplier.go (hunk 1 above)
Comment (bryanchriswhite) [reply testutil/keeper/supplier.go]: I've refactored to not need a new type. It seems like that was mostly what you were getting at.

File: testutil/testclient/localnet.go [hunk 13]
```diff
@@ -57,9 +61,13 @@ func NewLocalnetClientCtx(t *testing.T, flagSet *pflag.FlagSet) *client.Context
...
 // - A flag set populated with flags tailored for localnet environments.
-func NewLocalnetFlagSet(t *testing.T) *pflag.FlagSet {
+func NewLocalnetFlagSet(t gocuke.TestingT) *pflag.FlagSet {
+	t.Helper()
+
 	mockFlagSet := pflag.NewFlagSet("test", pflag.ContinueOnError)
-	mockFlagSet.String(flags.FlagNode, "tcp://127.0.0.1:36657", "use localnet poktrolld node")
+	// TODO_IMPROVE: It would be nice if the value could be set correctly based
+	// on whether the test using it is running in tilt or not.
+	mockFlagSet.String(flags.FlagNode, "tcp://sequencer-poktroll-sequencer:36657", "use localnet poktrolld node")
```
Comment (bryanchriswhite) [inline testutil/testclient/localnet.go:70]: @okdas (same as https://github.com/pokt-network/poktroll/pull/236#discussion_r1413781028)

File: pkg/client/tx/client.go [hunk 14]
```diff
@@ -89,7 +89,11 @@ type (
 // the transactions subscription.
 type TxEvent struct {
 	// Tx is the binary representation of the tx hash.
-	Tx     []byte            `json:"tx"`
+	Tx     []byte   `json:"tx"`
+	Result TxResult `json:"result"`
```
Comment (Olshansk) [inline pkg/client/tx/client.go:93]:
```suggestion
	// TxResult is the CommetBFT result of committing a transaction on-chain
	Result TxResult `json:"result"`
```

File: x/supplier/keeper/msg_server_create_claim.go [hunk 15]
```diff
@@ -45,11 +55,22 @@ func (k msgServer) CreateClaim(goCtx context.Context, msg *suppliertypes.MsgCrea
 	if !found {
 		return nil, suppliertypes.ErrSupplierNotFound.Wrapf(
 			"supplier address %q in session ID %q",
-			msg.SupplierAddress,
+			msg.GetSupplierAddress(),
 			sessionRes.GetSession().GetSessionId(),
 		)
 	}
 
+	// TODO_TECHDEBT(#181): refactor once structured logging is available.
```
Comment (Olshansk) [inline x/supplier/keeper/msg_server_create_claim.go:63]: Just a heads up that it'll be a couple rounds (because others have PRs in flight as well)

File: x/supplier/keeper/msg_server_create_claim.go [hunk 16]
```diff
@@ -64,14 +85,21 @@ func (k msgServer) CreateClaim(goCtx context.Context, msg *suppliertypes.MsgCrea
...
-	logger.Info("created claim for supplier %s at sessionRes ending height %d", claim.SupplierAddress, claim.SessionEndBlockHeight)
+	logger.
+		With(
+			"session_id", claim.GetSessionId(),
+			"session_end_height", claim.GetSessionEndBlockHeight(),
+			"supplier", claim.GetSupplierAddress(),
+		).
+		Debug("created claim")
 
+	// TODO_CONSIDERATION: perhaps it would be useful to return the claim in the response.
```
Comment (Olshansk) [inline x/supplier/keeper/msg_server_create_claim.go:103]:
```suggestion
	// TODO: return the claim in the response.
```

File: x/supplier/client/cli/helpers_test.go [hunk 17]
```diff
@@ -2,17 +2,39 @@
...
+	"encoding/json"
+	"fmt"
 	"strconv"
 	"testing"
 
+	"cosmossdk.io/math"
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/cosmos/cosmos-sdk/codec"
+	types3 "github.com/cosmos/cosmos-sdk/codec/types"
```
Comment (Olshansk) [inline x/supplier/client/cli/helpers_test.go:16]: NIT: Introduce semantically named types

File: x/supplier/client/cli/helpers_test.go [hunk 18]
```diff
@@ -32,3 +54,197 @@ func networkWithSupplierObjects(t *testing.T, n int) (*network.Network, []shared
...
+// into something more generic and moved into a shared testutil package.
+// TODO_TECHDEBT: refactor; this function has more than a single responsibility,
+// which should be to configure and start the test network. The genesis state,
+// accounts, and claims set up logic can probably be factored out and/or reduced.
+func networkWithClaimObjects(
+	t *testing.T,
+	sessionCount int,
+	supplierCount int,
+	appCount int,
+	// TODO_THIS_COMMIT: hook up to genesis state generation...
```
Comment (Olshansk) [inline x/supplier/client/cli/helpers_test.go:68]: TODO_IN_THIS_COMMIT. 

Still intending todo in this commit?

File: testutil/testclient/localnet.go [hunk 19]
```diff
@@ -14,7 +13,10 @@ import (
 )
 
 // CometLocalWebsocketURL provides a default URL pointing to the localnet websocket endpoint.
-const CometLocalWebsocketURL = "ws://localhost:36657/websocket"
+//
+// TODO_IMPROVE: It would be nice if the value could be set correctly based
+// on whether the test using it is running in tilt or not.
+const CometLocalWebsocketURL = "ws://sequencer-poktroll-sequencer:36657/websocket"
```
[the remark this answers is not in the document]
Comment (okdas) [reply testutil/testclient/localnet.go:19]: @bryanchriswhite yep! That hostname will now resolve to sequencer on both LocalNet and DevNet!

File: x/supplier/client/cli/helpers_test.go [hunk 20]
```diff
@@ -32,3 +54,197 @@ func networkWithSupplierObjects(t *testing.T, n int) (*network.Network, []shared
...
+				// TODO_TECHDEBT(#196): Move this outside of the forloop so that the test iteration is faster
+				require.NoError(t, net.WaitForNextBlock())
+			}
+		}
+	}
+
+	return net, claims
+}
+
+func encodeSessionHeader(
```
Comment (Olshansk) [inline x/supplier/client/cli/helpers_test.go:162]:
```suggestion
// encodeSessionHeader returns a base64 encoded string of a json
// serialized session header.
func encodeSessionHeader(
```

File: x/supplier/client/cli/helpers_test.go [hunk 21]
```diff
@@ -32,3 +54,197 @@ func networkWithSupplierObjects(t *testing.T, n int) (*network.Network, []shared
...
+	}
+
+	responseRaw, err := cli.ExecTestCLICmd(ctx, cli2.CmdCreateClaim(), args)
+	require.NoError(t, err)
+	var responseJson map[string]interface{}
+	err = json.Unmarshal(responseRaw.Bytes(), &responseJson)
+	require.NoError(t, err)
+	require.Equal(t, float64(0), responseJson["code"], "code is not 0 in the response: %v", responseJson)
+
+	return &types.Claim{
```
Comment (Olshansk) [inline x/supplier/client/cli/helpers_test.go:214]:
```suggestion
	// TODO_TECHDEBT:  Forward the actual claim in the response once the response is updated to return it.
	return &types.Claim{
```

File: x/supplier/client/cli/helpers_test.go [hunk 22]
```diff
@@ -32,3 +54,197 @@ func networkWithSupplierObjects(t *testing.T, n int) (*network.Network, []shared
...
+		SessionId:               sessionId,
+		SessionEndBlockHeight:   sessionStartHeight + numBlocksPerSession,
+		Service:                 &sharedtypes.Service{Id: testServiceId},
+	}
+	cdc := codec.NewProtoCodec(types3.NewInterfaceRegistry())
+	sessionHeaderBz := cdc.MustMarshalJSON(argSessionHeader)
+	return base64.StdEncoding.EncodeToString(sessionHeaderBz)
+}
+
+func createClaim(
```
Comment (Olshansk) [inline x/supplier/client/cli/helpers_test.go:182]:
```suggestion
// createClaim sends a tx using the test CLI to create an on-chain claim
func createClaim(
```

File: testutil/testclient/localnet.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply testutil/testclient/localnet.go:19]: @okdas Bump

File: pkg/client/tx/client.go [hunk 23]
```diff
@@ -89,7 +89,11 @@ type (
...
+}
+
+type TxResult struct {
```
Comment (Olshansk) [inline pkg/client/tx/client.go:96]: I'm worried that `TxResult` may be conflated with the definition of a TxResult in Comet/Tendermint land. Wdyt if `ABCIEvents`, `TxResultEvents` or removing the decoupling altogether?

Screenshot 2023-12-04 at 3 07 34 PM

https://github.com/cosmos/cosmos-sdk/blob/main/proto/tendermint/abci/types.proto#L416

File: e2e/tests/session_steps_test.go [hunk 24]
```diff
@@ -79,60 +83,65 @@ func (s *suite) AfterTheSupplierCreatesAClaimForTheSession() {
...
-	// TODO_IN_THIS_COMMIT: query claims before this step, perhaps note the highest
-	// session end height, then compare against queried claims in this step,
-	// asserting the length is +1 and the highest session end height increased.
-	//require.Lenf(s, claimsRes.Claim, 1, "expected 1 claim, got %d", len(claimsRes.Claim))
-	require.NotEmpty(s, claimsRes.Claim)
+	// Assert that the number of claims has increased by one.
+	preExistingClaims := s.scenarioState[preExistingClaimsKey].([]suppliertypes.Claim)
+	require.Len(s, claimsRes.Claim, len(preExistingClaims)+1)
+
+	// TODO_IMPROVE: assert that the root hash of the claim contains the correct
```
Comment (Olshansk) [inline e2e/tests/session_steps_test.go:101]: 👍

File: testutil/supplier/fixtures.go [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+
+// NewSessionFixturesWithPairings creates a map of session fixtures where the key
+// is the application address and the value is the session fixture. App/supplier
+// addresses are expected to be provided in alternating order (as pairs). The same
+// app and/or supplier may be given more than once but only distinct pairs will
+// be added to the session fixtures map.
+func NewSessionFixturesWithPairings(
+	t *testing.T,
+	service *sharedtypes.Service,
+	appAndSupplierAddrPairs ...string,
```
Comment (Olshansk) [inline testutil/supplier/fixtures.go:32]: I think this warrants a helper struct. It'll make the business logic cleaner. 

If you want to get this over the finish line though, I'd be okay with a `TODO_TECHDEBT/TODO_COMMUNITY`

File: testutil/supplier/fixtures.go [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+			Suppliers: []*sharedtypes.Supplier{
+				newSupplier(t, supplierAddr, service),
+			},
+		}
+	}
+
+	return sessionFixturesByAppAddr
+}
+
+func newSupplier(t *testing.T, addr string, services ...*sharedtypes.Service) *sharedtypes.Supplier {
```
Comment (Olshansk) [inline testutil/supplier/fixtures.go:78]:
```suggestion
// newSuppliers configures a supplier for the services provided and nil endpoints.
func newSupplier(t *testing.T, addr string, services ...*sharedtypes.Service) *sharedtypes.Supplier {
```

File: testutil/supplier/fixtures.go [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	return &sharedtypes.Supplier{
+		Address:  addr,
+		Stake:    &sdktypes.Coin{},
+		Services: serviceConfigs,
+	}
+}
+
+func newApplication(t *testing.T, addr string, services ...*sharedtypes.Service) *apptypes.Application {
```
Comment (Olshansk) [inline testutil/supplier/fixtures.go:96]:
```suggestion
// newApplication configures a supplier for the services provided.
func newApplication(t *testing.T, addr string, services ...*sharedtypes.Service) *apptypes.Application {
```

File: x/supplier/client/cli/helpers_test.go [hunk 28]
```diff
@@ -32,3 +54,197 @@ func networkWithSupplierObjects(t *testing.T, n int) (*network.Network, []shared
...
+
+	return &types.Claim{
+		SupplierAddress:       supplierAddr,
+		SessionId:             sessionId,
+		SessionEndBlockHeight: uint64(sessionEndHeight),
+		RootHash:              rootHash,
+	}
+}
+
+func getSessionId(
```
Comment (Olshansk) [inline x/supplier/client/cli/helpers_test.go:222]:
```suggestion
// getSessionId sends a query using the test CLI to get a session for the inputs provided.
// It is assumed that the supplierAddr will be in that session based on the test design, but this
// is insured in this function before it's successfully returned.
func getSessionId(
```

File: x/supplier/client/cli/helpers_test.go [hunk 29]
```diff
@@ -32,3 +54,197 @@ func networkWithSupplierObjects(t *testing.T, n int) (*network.Network, []shared
...
+	supplierAddr string,
+	sessionStartHeight int64,
+) string {
+	t.Helper()
+	ctx := context.TODO()
+
+	sessionQueryClient := types2.NewQueryClient(net.Validators[0].ClientCtx)
+	res, err := sessionQueryClient.GetSession(ctx, &types2.QueryGetSessionRequest{
+		ApplicationAddress: appAddr,
+		Service:            &sharedtypes.Service{Id: testServiceId},
```
Comment (Olshansk) [inline x/supplier/client/cli/helpers_test.go:235]: Thoughts on having the caller pass in this local constant?

**Review by Olshansk — CHANGES_REQUESTED** [review]
The PR is looking really good overall. Really appreciate both the changes, and comments to set up followup work.

Left a few lingering comments but should be g2g right after!

File: pkg/client/tx/client.go (hunk 14 above)
Comment (bryanchriswhite) [reply pkg/client/tx/client.go:93]: Tyty :pray:

File: x/supplier/client/cli/helpers_test.go (hunk 17 above)
Comment (bryanchriswhite) [reply x/supplier/client/cli/helpers_test.go:16]: Nice catch, this was probably my IDE doing something dumb when I refactored some things.

File: x/supplier/client/cli/helpers_test.go (hunk 18 above)
Comment (bryanchriswhite) [reply x/supplier/client/cli/helpers_test.go:68]: I'm gonna remove the comment and the unused `serviceCount` param. My original thinking was that it would make sense for this function to parameterize services as well, but the need hasn't quite arisen yet so I would put it off at this point.

File: x/supplier/client/cli/helpers_test.go (hunk 29 above)
Comment (bryanchriswhite) [reply x/supplier/client/cli/helpers_test.go:235]: Definitely have thoughts :sweat_smile:. This is related to https://github.com/pokt-network/poktroll/pull/236#discussion_r1415261843. 

TL;DR, I would  hold off for now.

It might be better to take a step back and think about how we would like multi-service integration tests to be designed, and then evaluate if it's compatible with our current direction (i.e. refactor and/or reuse what we have) or if we need to do that in a new test suite with more comprehensive helpers. Wdyt?

File: pkg/client/tx/client.go (hunk 23 above)
Comment (bryanchriswhite) [reply pkg/client/tx/client.go:97]: Thanks for calling this out! :raised_hands:  This is indeed the correct type and we should be using it instead of the one(s) currently defined in our tx client package. :mechanical_arm: 

I have gone digging (on more than one occasion) for the actual protobuf that is being used by comet/cosmos to produce the JSON structure we see in the event subscription stream. I can't remember if I looked at this one before or not but I'm realizing that it's tricky to connect the dots because it's being serialized to JSON. :thinking:  Anyways, you nailed it - thanks!

File: testutil/supplier/fixtures.go (hunk 25 above)
Comment (bryanchriswhite) [reply testutil/supplier/fixtures.go:32]: Good call. I found a  low-verbosity solution. :+1:

File: pkg/client/tx/client.go [hunk 30]
```diff
@@ -87,15 +87,10 @@ type (
...
-type TxEvent struct {
-	// Tx is the binary representation of the tx hash.
-	Tx     []byte   `json:"tx"`
-	Result TxResult `json:"result"`
-}
-
-type TxResult struct {
-	Events []abciTypes.Event `json:"events"`
-}
+//
```
Comment (Olshansk) [inline pkg/client/tx/client.go:90]: So much cleaner!

File: testutil/supplier/fixtures.go [hunk 31]
```diff
@@ -21,6 +21,12 @@ const (
 // application address and the value is the session fixture.
 type SessionsByAppAddress map[string]sessiontypes.Session
 
+// AppSupplierPair is a pairing of an application and a supplier address.
```
Comment (Olshansk) [inline testutil/supplier/fixtures.go:24]: Ty for making the change!

File: x/supplier/client/cli/helpers_test.go (hunk 29 above)
Comment (Olshansk) [reply x/supplier/client/cli/helpers_test.go:240]: I think holding off and seeing what/when we need from multi-service integration makes sense.

Let's hold off 👍

**Review by Olshansk — APPROVED** [review]
LGTM. The last round of comments & cleanup really put the :bowtie: on it!

Note that the E2E tests are failing and wanted to make sure you remember the base branch here is not `main` (should it be?), but otherwise g2g!

**Comment by bryanchriswhite** [comment]
> Note that the E2E tests are failing and wanted to make sure you remember the base branch here is not `main` (should it be?), but otherwise g2g!

Thanks for calling this out! I typically intentionally keep the github base of PR branches the same as the git (actual) base until after approval to avoid cluttering the commit list during review. I always take note of the +/- LOC diff before and after re-github-basing (not an actual rebase operation). In the event of discrepancies and/or merge conflict resolutions I double-check everything before merging and would ask for re-review if something unexpected arises or changes since approval.

File: x/supplier/client/cli/helpers_test.go (hunk 29 above)
Comment (bryanchriswhite) [reply x/supplier/client/cli/helpers_test.go:240]: :pray:
